pkg/sql/plan/function/builtin/unary/time_of_day.go

```diff
@@ -0,0 +1,201 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unary
+
+import (
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+	"github.com/matrixorigin/matrixone/pkg/container/vector"
+	"github.com/matrixorigin/matrixone/pkg/vm/process"
+)
+
+func DatetimeToHour(vectors []*vector.Vector, proc *process.Process) (*vector.Vector, error) {
+	inputVector := vectors[0]
+	resultType := types.New(types.T_uint8, 0, 0, 0)
+	inputValues := vector.MustTCols[types.Datetime](inputVector)
+	if inputVector.IsScalar() {
+		if inputVector.IsScalarNull() {
+			return proc.AllocScalarNullVector(resultType), nil
+		}
+		resultVector := vector.NewConst(resultType, 1)
+		vector.SetCol(resultVector, []uint8{uint8(inputValues[0].Hour())})
+		return resultVector, nil
+	} else {
+		resultVector := vector.New(resultType)
+		for i, v := range inputValues {
+			if inputVector.GetNulls().Contains(uint64(i)) {
+				resultVector.GetNulls().Set(uint64(i))
+				if err := resultVector.Append(uint8(0), true, proc.GetMheap()); err != nil {
+					return nil, err
+				}
+				continue
+			}
+			if err := resultVector.Append(uint8(v.Hour()), false, proc.GetMheap()); err != nil {
+				return nil, err
+			}
+		}
+		return resultVector, nil
+	}
+}
+
+func TimestampToHour(vectors []*vector.Vector, proc *process.Process) (*vector.Vector, error) {
+	inputVector := vectors[0]
+	resultType := types.New(types.T_uint8, 0, 0, 0)
+	inputValues := vector.MustTCols[types.Timestamp](inputVector)
+	tmp := make([]types.Datetime, len(inputValues))
+	types.TimestampToDatetime(proc.SessionInfo.TimeZone, inputValues, tmp)
+	if inputVector.IsScalar() {
+		if inputVector.IsScalarNull() {
+			return proc.AllocScalarNullVector(resultType), nil
+		}
+		resultVector := vector.NewConst(resultType, 1)
+		vector.SetCol(resultVector, []uint8{uint8(tmp[0].Hour())})
+		return resultVector, nil
+	} else {
+		resultVector := vector.New(resultType)
+		for i, v := range tmp {
+			if inputVector.GetNulls().Contains(uint64(i)) {
+				resultVector.GetNulls().Set(uint64(i))
+				if err := resultVector.Append(uint8(0), true, proc.GetMheap()); err != nil {
+					return nil, err
+				}
+				continue
+			}
+			if err := resultVector.Append(uint8(v.Hour()), false, proc.GetMheap()); err != nil {
+				return nil, err
+			}
+		}
+		return resultVector, nil
+	}
+}
+
+func DatetimeToMinute(vectors []*vector.Vector, proc *process.Process) (*vector.Vector, error) {
+	inputVector := vectors[0]
+	resultType := types.New(types.T_uint8, 0, 0, 0)
+	inputValues := vector.MustTCols[types.Datetime](inputVector)
+	if inputVector.IsScalar() {
+		if inputVector.IsScalarNull() {
+			return proc.AllocScalarNullVector(resultType), nil
+		}
+		resultVector := vector.NewConst(resultType, 1)
+		vector.SetCol(resultVector, []uint8{uint8(inputValues[0].Minute())})
+		return resultVector, nil
+	} else {
+		resultVector := vector.New(resultType)
+		for i, v := range inputValues {
```

Comment:
you can check input has any null out of for loops。

if nulls.Any(inputVector.Nsp) {
    // need check if item is null
} else {
   // just append data to vector without null check
}

Reply:
> you can check if input has any null out of for loops。
> 
> if nulls.Any(inputVector.Nsp) { // need check if item is null } else { // just append data to vector without null check }

I did a simple benchmark to compare the performance between checking in/out `for` loop. There is little difference in performance between these two approaches, therefore I think the origin is more simple and save lines of code.

origin

<img width="658" alt="image" src="https://user-images.githubusercontent.com/77312370/192279727-e56e5861-e571-443c-a818-141eef5c91ee.png">

check input has any null out of for loops
<img width="664" alt="image" src="https://user-images.githubusercontent.com/77312370/192279951-7ab3850e-01a2-4ab1-98e9-16feda386469.png">